src/preferences/common/compound-preference-server.ts

```diff
@@ -5,45 +5,52 @@
  * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
  */
 
-import { PreferenceServer, PreferenceClient } from './preference-protocol'
+import { PreferenceServer, PreferenceClient, PreferenceChangedEvent } from './preference-protocol'
+import { DefaultPreferenceServer } from './default-preference-server'
 
 export class CompoundPreferenceServer implements PreferenceServer {
 
     protected readonly servers: PreferenceServer[];
+    protected client: PreferenceClient | undefined;
+    protected serversReady: boolean = false;
+
     constructor(
-        ...servers: PreferenceServer[]
+        defaultServer: DefaultPreferenceServer, ...servers: PreferenceServer[],
     ) {
         this.servers = servers;
-    }
-
-    dispose(): void {
-        for (const server of this.servers) {
-            server.dispose();
+        for (const server of servers) {
+            server.setClient({
+                onDidChangePreference: event => this.onDidChangePreference(event)
+            })
         }
     }
 
-    async has(preferenceName: string): Promise<boolean> {
-        for (const server of this.servers) {
-            if (await server.has(preferenceName)) {
-                return true;
+    // TODO scope management should happen here
+    protected onDidChangePreference(event: PreferenceChangedEvent): void {
+        if (event.preferenceName == "INIT") {
```

Comment:
In that case, I think I prefer the first way i.e receive the first message with all preferences rather than an explicit request.